akka-stream/src/main/scala/akka/stream/impl/RetryFlowCoordinator.scala

```diff
@@ -0,0 +1,178 @@
+    parallelism: Long,
+    retryWith: PartialFunction[(Try[Out], State), Option[immutable.Iterable[(In, State)]]],
+    minBackoff: FiniteDuration,
+    maxBackoff: FiniteDuration,
+    randomFactor: Double)
+    extends GraphStage[
+      BidiShape[(In, State), (Try[Out], State), (Try[Out], State, InternalState), (In, State, InternalState)]] {
+
+  private val externalIn = Inlet[(In, State)]("RetryFlow.externalIn")
+  private val externalOut = Outlet[(Try[Out], State)]("RetryFlow.externalOut")
+
+  private val internalIn = Inlet[(Try[Out], State, InternalState)]("RetryFlow.internalIn")
+  private val internalOut = Outlet[(In, State, InternalState)]("RetryFlow.internalOut")
+
+  override val shape
+      : BidiShape[(In, State), (Try[Out], State), (Try[Out], State, InternalState), (In, State, InternalState)] =
+    BidiShape(externalIn, externalOut, internalIn, internalOut)
+
+  override def createLogic(attributes: Attributes): GraphStageLogic = new TimerGraphStageLogic(shape) {
+    private var numElementsInCycle = 0
+    private var queueRetries =
+      scala.collection.immutable.SortedSet.empty[(In, State, InternalState)](Ordering.fromLessThan { (e1, e2) =>
+        if (e1._3.retryDeadline != e2._3.retryDeadline) e1._3.retryDeadline < e2._3.retryDeadline
+        else e1.hashCode < e2.hashCode
+      })
+    private val queueOut = scala.collection.mutable.Queue.empty[(Try[Out], State)]
+
+    private final val RetryTimer = "retry_timer"
+
+    setHandler(
+      externalIn,
+      new InHandler {
+        override def onPush(): Unit = {
+          val is = {
+            val in = grab(externalIn)
+            (in._1, in._2, InternalState())
+          }
+          if (isAvailable(internalOut)) {
+            push(internalOut, is)
+            numElementsInCycle += 1
+          } else {
+            queueRetries += is
+          }
+        }
+
+        override def onUpstreamFinish(): Unit =
+          if (numElementsInCycle == 0 && queueRetries.isEmpty) {
+            if (queueOut.isEmpty) completeStage()
+            else emitMultiple(externalOut, queueOut.iterator, () => completeStage())
+          }
+      })
+
+    setHandler(
+      externalOut,
+      new OutHandler {
+        override def onPull(): Unit =
+          // prioritize pushing queued element if available
+          if (queueOut.isEmpty) {
+            if (!hasBeenPulled(internalIn)) pull(internalIn)
+          } else push(externalOut, queueOut.dequeue())
+      })
+
+    setHandler(
+      internalIn,
+      new InHandler {
+        override def onPush(): Unit = {
+          numElementsInCycle -= 1
+          grab(internalIn) match {
+            case (out, s, internalState) =>
+              retryWith
+                .applyOrElse[(Try[Out], State), Option[immutable.Iterable[(In, State)]]]((out, s), _ => None)
+                .fold(pushAndCompleteIfLast((out, s))) {
+                  xs =>
+                    val current = System.currentTimeMillis()
```

Comment:
I'd prefer that we used System.nanoTime (for example because System.currentTimeMillis may jump backwards),
but I see that queueRetries is sorted on the absolute value of the timestamp and that will not work for nanoTime.

One way to solve that is to capture a nonoTime offset when the stage is created, and then sort on the diff between the nanoTime value in InternalState and that offset. Maybe that diff can be the value in InternalState if it's not used for anything else.